sql/src/main/java/io/crate/metadata/doc/DocTableInfo.java

```diff
@@ -209,38 +216,60 @@ private Routing getRouting(
                     missingShards.add(shardIterator.shardId());
                 }
             } else {
-                throw new UnavailableShardsException(shardIterator.shardId());
+                if (isPartitioned) {
+                    // if the table is partitioned maybe a index/shard just got newly created ...
+                    missingShards.add(shardIterator.shardId());
+                } else {
+                    throw new UnavailableShardsException(shardIterator.shardId());
+                }
             }
         }
         if (missingShards.isEmpty()) {
             return new Routing(locations);
         } else {
+            if (!retry) {
+                throw new UnavailableShardsException(missingShards.get(0));
+            } else if (currentRetry > MAX_ROUTING_RETRIES) {
+                throw new UnavailableShardsException(missingShards.get(0));
+            }
             final SettableFuture<Routing> futureRouting = SettableFuture.create();
             observer.waitForNextChange(new ClusterStateObserver.Listener() {
+
+                Future<?> routingFuture;
+
                 @Override
                 public void onNewClusterState(ClusterState state) {
-                    Thread thread = new Thread(new Runnable() {
-                        @Override
-                        public void run() {
-                            try {
-                                futureRouting.set(getRouting(observer, whereClause, preference));
-                            } catch (Throwable e) {
-                                futureRouting.setException(e);
+                    try {
+                        routingFuture = executorService.submit(new Runnable() {
```

Comment:
maybe filter with a predicate to only include new ClusterStates which have a changed routing.